File: 2327-number-of-people-aware-of-a-secret/2327. Number of People Aware of a Secret.java
class Solution {

    private final static long MOD = 1_000_000_000 + 7;

    public int peopleAwareOfSecret(int n, int delay, int forget) {
        final long[] spread = new long[n + 1];        
        final long[] stop = new long[n + 1];

        if(1 + delay <= n) {
            spread[1 + delay] = 1;
        }
        if(1 + forget <= n) {
            stop[1 + forget] = 1;
        }

        long peopleCount = 1;
        long newPeopleIncomingCount = 0;

        for(int i = 2; i <= n; ++i) {
            newPeopleIncomingCount -= stop[i] % MOD;
            newPeopleIncomingCount += spread[i] % MOD;

            peopleCount += newPeopleIncomingCount % MOD;
            peopleCount -= stop[i] % MOD;
            // System.out.println(i + "<>" + newPeopleIncomingCount + "<>" + peopleCount + "<>" + spread[i] + "<>" + stop[i]);

            final int spreadStart = i + delay;
            if(spreadStart <= n) {
                spread[spreadStart] = newPeopleIncomingCount % MOD;
            }

            final int spreadStop = i + forget;
            if(spreadStop <= n) {
                stop[spreadStop] = newPeopleIncomingCount % MOD;
            }
        }

        return (int) (peopleCount % MOD);
    }

}

File: 2327-number-of-people-aware-of-a-secret/2327. Number of People Aware of a Secret.cpp
// Brute Way -> O(n^2) TLE even with 
// class Solution {
// public:
//     int peopleAwareOfSecret(int n, int delay, int forget) {
//         int timer = 1, peeps = 0;
//         priority_queue< pair<int,int>, vector<pair<int,int>> , greater<pair<int,int>>>pq;
//         pq.push({1+forget, 1});
//         while (timer <= n){
//             vector<pair<int,int>>temp;
//             while (pq.top().first <= timer) pq.pop();
//             while (pq.top().second + delay <= timer && pq.top().first > timer){
//                 temp.push_back(pq.top());
//                 pq.pop();
//                 pq.push({timer+forget, timer});
//             }
//             for(auto &i:temp) pq.push(i);
//             if (timer == n) peeps = pq.size();
//             timer++;
//         }
//         return peeps;
//     }
// };
int MOD = 1e9+7;
class Solution {
public:
    int rec(int day, int delay, int forget, vector<int>&days){
        if (day == 1) return 1; // since only 1 person knew the secret at day 1
        if (days[day] != -1) return days[day]; 
        int total_people_who_can_share=0;
        for(int cur_day = day-forget+1; cur_day<=day-delay; cur_day++){
            if (cur_day>0) total_people_who_can_share = (total_people_who_can_share + rec(cur_day, delay, forget, days)) % MOD;
        }
        return days[day] = total_people_who_can_share;
    }
    int peopleAwareOfSecret(int n, int delay, int forget) {
        int total_peeps_on_dayn = 0;
        vector<int>days(n+1,-1); days[1]=1;
        for(int prev = n-forget+1; prev<=n; prev++){
            total_peeps_on_dayn = (total_peeps_on_dayn + rec(prev, delay, forget, days))%MOD;
        }
        return total_peeps_on_dayn;
    }
};